bootstrap/bootstrap.go

```diff
@@ -0,0 +1,317 @@
+	snaps := []string{}
+	snaps = append(snaps, y.Bootstrap.Snaps...)
+	snaps = append(snaps, model.Gadget())
+	snaps = append(snaps, model.OS())
+	snaps = append(snaps, model.Kernel())
+	snaps = append(snaps, model.RequiredSnaps()...)
+
+	for _, snapName := range snaps {
+		fmt.Printf("Fetching %s\n", snapName)
+		if osutil.FileExists(snapName) {
+			if err := copyLocalSnapFile(snapName); err != nil {
+				return err
+			}
+		} else {
+			if _, err := downloadSnapWithSideInfo(snapName, opts); err != nil {
+				return err
+			}
+		}
+	}
+
+	// now do the bootloader stuff
+	if err := setBootvars(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func setBootvars() error {
+	// set the bootvars for kernel/os snaps so that the system
+	// actually boots and can do the `firstboot` import of the snaps.
+	//
+	// there is also no mounted os/kernel snap, all we have are the
+	// blobs
+
+	bootloader, err := partition.FindBootloader()
+	if err != nil {
+		return fmt.Errorf("can not set kernel/os bootvars: %s", err)
+	}
+
+	snaps, _ := filepath.Glob(filepath.Join(dirs.SnapBlobDir, "*.snap"))
+	if len(snaps) == 0 {
+		return fmt.Errorf("internal error: cannot find os/kernel snap")
+	}
+	for _, fullname := range snaps {
+		bootvar := ""
+		bootvar2 := ""
+
+		// detect type
+		snapFile, err := snap.Open(fullname)
+		if err != nil {
+			return fmt.Errorf("can not read %v", fullname)
+		}
+		// read .sideinfo
+		var si snap.SideInfo
+		siFn := fullname + ".sideinfo"
+		if osutil.FileExists(siFn) {
+			j, err := ioutil.ReadFile(siFn)
+			if err != nil {
+				return err
+			}
+			if err := json.Unmarshal(j, &si); err != nil {
+				return fmt.Errorf("cannot read metadata: %s %s\n", siFn, err)
+			}
+		}
+		info, err := snap.ReadInfoFromSnapFile(snapFile, &si)
+		if err != nil {
+			return fmt.Errorf("can not get info for %v", fullname)
+		}
+		// local install
+		if info.Revision.Unset() {
+			info.Revision = snap.R(-1)
+		}
+
+		switch info.Type {
+		case snap.TypeOS:
+			bootvar = "snappy_os"
+			bootvar2 = "snappy_good_os"
+		case snap.TypeKernel:
+			bootvar = "snappy_kernel"
+			bootvar2 = "snappy_good_kernel"
+			if err := extractKernelAssets(fullname, info); err != nil {
+				return err
+			}
+		}
+
+		name := filepath.Base(fullname)
+		for _, b := range []string{bootvar, bootvar2} {
+			if b != "" {
+				if err := bootloader.SetBootVar(b, name); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func runCommand(cmdStr ...string) error {
+	cmd := exec.Command(cmdStr[0], cmdStr[1:]...)
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("cannot run %q: %s (%s)", cmd, err, output)
+	}
+	return nil
+}
+
+func extractKernelAssets(snapPath string, info *snap.Info) error {
+	// FIXME: hrm, hrm, we need to be root for this - alternatively
+	//        we could make boot.ExtractKernelAssets() work on
+	//        a plain .snap file again (i.e. revert bee59a2)
```

Comment:
it's my fault I didn't think of this use case, now that we have dir containers though there is probably an elegant solution that is not fully reverting but reading out of the squashfs blob only if needed